File: RealtimeFeed/Sources/RealtimeFeed/PriceFeedClient.swift
import Foundation

/// Connects to Binance's combined WebSocket stream, resubscribes on every reconnect,
/// and exposes a conflated snapshot stream so a slow UI consumer never falls behind
/// a high-message-rate feed — it just gets dropped intermediate snapshots and always
/// renders the latest known price per symbol.
public actor PriceFeedClient {
    public let symbols: [String]

    private let session: URLSession
    // Subscriptions are encoded directly in the URL (Binance's "combined streams" form) rather
    // than sent as a control frame after connecting — sending immediately after `resume()` races
    // the WebSocket handshake (URLSessionWebSocketTask gives no "now connected" signal without a
    // delegate) and fails with "Socket is not connected". Encoding subscriptions in the URL also
    // makes resubscribe-on-reconnect trivial: reconnecting to the same URL *is* resubscribing.
    private let endpoint: URL
    private let reconnectPolicy: ReconnectPolicy

    private var task: URLSessionWebSocketTask?
    private var isRunning = false
    private var reconnectAttempt = 0
    private var latestBySymbol: [String: PriceTick] = [:]

    private var stateContinuation: AsyncStream<ConnectionState>.Continuation?
    private var snapshotContinuation: AsyncStream<[String: PriceTick]>.Continuation?

    // `nonisolated` because AsyncStream<T> is Sendable and these are set once in init —
    // callers need to iterate them from outside the actor without hopping isolation.
    public nonisolated let states: AsyncStream<ConnectionState>
    public nonisolated let snapshots: AsyncStream<[String: PriceTick]>

    public init(symbols: [String], reconnectPolicy: ReconnectPolicy = ReconnectPolicy(), session: URLSession = URLSession(configuration: .default)) {
        let lowercased = symbols.map { $0.lowercased() }
        self.symbols = lowercased
        self.reconnectPolicy = reconnectPolicy
        self.session = session

        let streams = lowercased.map { "\($0)@trade" }.joined(separator: "/")
        self.endpoint = URL(string: "wss://stream.binance.com:9443/stream?streams=\(streams)")!

        var stateCont: AsyncStream<ConnectionState>.Continuation!
        self.states = AsyncStream(bufferingPolicy: .bufferingNewest(1)) { stateCont = $0 }
        self.stateContinuation = stateCont

        var snapCont: AsyncStream<[String: PriceTick]>.Continuation!
        self.snapshots = AsyncStream(bufferingPolicy: .bufferingNewest(1)) { snapCont = $0 }
        self.snapshotContinuation = snapCont
    }

    public func start() {
        guard !isRunning else { return }
        isRunning = true
        Task { await runLoop() }
    }

    public func stop() {
        isRunning = false
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        stateContinuation?.yield(.disconnected)
    }

    private func runLoop() async {
        while isRunning {
            do {
                try await connectAndSubscribe()
                reconnectAttempt = 0
                try await receiveUntilFailure()
            } catch {
                guard isRunning else { break }
                FileHandle.standardError.write(Data("[PriceFeedClient] error: \(error)\n".utf8))
                reconnectAttempt += 1
                let delay = reconnectPolicy.delay(forAttempt: reconnectAttempt)
                stateContinuation?.yield(.reconnecting(attempt: reconnectAttempt, delay: delay))
                try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            }
        }
        stateContinuation?.yield(.disconnected)
    }

    private func connectAndSubscribe() async throws {
        stateContinuation?.yield(reconnectAttempt == 0 ? .connecting : .reconnecting(attempt: reconnectAttempt, delay: 0))

        let webSocketTask = session.webSocketTask(with: endpoint)
        webSocketTask.resume()
        self.task = webSocketTask

        stateContinuation?.yield(.connected)
    }

    private func receiveUntilFailure() async throws {
        while isRunning, let task {
            let message = try await task.receive()
            switch message {
            case .string(let text):
                handle(text: text)
            case .data(let raw):
                if let text = String(data: raw, encoding: .utf8) {
                    handle(text: text)
                }
            @unknown default:
                break
            }
        }
    }

    private func handle(text: String) {
        guard let tick = Self.parseTick(json: text) else { return }
        latestBySymbol[tick.symbol] = tick
        snapshotContinuation?.yield(latestBySymbol)
    }

    static func parseTick(json: String) -> PriceTick? {
        guard let data = json.data(using: .utf8),
              let envelope = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let payload = envelope["data"] as? [String: Any],
              let priceString = payload["p"] as? String,
              let price = Double(priceString),
              let symbol = payload["s"] as? String
        else { return nil }
        return PriceTick(symbol: symbol, price: price, timestamp: Date())
    }
}

File: shared/src/commonMain/kotlin/dev/web3demo/realtimefeed/PriceFeedClient.kt
package dev.web3demo.realtimefeed

import io.ktor.client.HttpClient
import io.ktor.client.plugins.websocket.WebSockets
import io.ktor.client.plugins.websocket.webSocket
import io.ktor.websocket.Frame
import io.ktor.websocket.readText
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.datetime.Clock
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Connects to Binance's combined WebSocket stream, mirroring RealtimeFeed/PriceFeedClient.swift
 * so both platforms share the same reconnect/backpressure design even before any UI exists.
 *
 * Subscriptions are encoded directly in the URL (`?streams=...`) rather than sent as a control
 * frame after connecting, same reasoning as the Swift version: sending immediately after the
 * socket opens races the handshake. Reconnecting to the same URL *is* resubscribing.
 *
 * `snapshots` is a StateFlow, which conflates by construction — a slow collector only ever sees
 * the latest full snapshot, never falls behind, and never needs manual buffering-policy tuning
 * the way the Swift AsyncStream side does explicitly.
 */
class PriceFeedClient(
    symbols: List<String>,
    private val reconnectPolicy: ReconnectPolicy = ReconnectPolicy(),
    private val client: HttpClient = HttpClient { install(WebSockets) },
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
) {
    private val streamsParam = symbols.joinToString("/") { "${it.lowercase()}@trade" }
    private val url = "wss://stream.binance.com:9443/stream?streams=$streamsParam"

    private val _state = MutableStateFlow<ConnectionState>(ConnectionState.Disconnected)
    val state: StateFlow<ConnectionState> = _state.asStateFlow()

    private val _snapshots = MutableStateFlow<Map<String, PriceTick>>(emptyMap())
    val snapshots: StateFlow<Map<String, PriceTick>> = _snapshots.asStateFlow()

    private var job: Job? = null
    private var reconnectAttempt = 0

    fun start() {
        if (job?.isActive == true) return
        job = scope.launch { runLoop() }
    }

    fun stop() {
        job?.cancel()
        job = null
        _state.value = ConnectionState.Disconnected
    }

    private suspend fun runLoop() {
        while (currentCoroutineContext().isActive) {
            try {
                _state.value =
                    if (reconnectAttempt == 0) {
                        ConnectionState.Connecting
                    } else {
                        ConnectionState.Reconnecting(reconnectAttempt, 0)
                    }

                client.webSocket(urlString = url) {
                    _state.value = ConnectionState.Connected
                    reconnectAttempt = 0
                    for (frame in incoming) {
                        if (frame is Frame.Text) handle(frame.readText())
                    }
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                reconnectAttempt += 1
                val delayMs = reconnectPolicy.delayMillis(reconnectAttempt)
                _state.value = ConnectionState.Reconnecting(reconnectAttempt, delayMs)
                delay(delayMs)
            }
        }
    }

    private fun handle(text: String) {
        val tick = parseTick(text) ?: return
        _snapshots.value = _snapshots.value + (tick.symbol to tick)
    }

    companion object {
        fun parseTick(json: String): PriceTick? =
            try {
                val envelope = Json.parseToJsonElement(json).jsonObject
                val data = envelope["data"]?.jsonObject ?: return null
                val price = data["p"]?.jsonPrimitive?.content?.toDoubleOrNull() ?: return null
                val symbol = data["s"]?.jsonPrimitive?.content ?: return null
                PriceTick(symbol = symbol, price = price, timestampMillis = Clock.System.now().toEpochMilliseconds())
            } catch (e: Exception) {
                null
            }
    }
}
